### PR TITLE
Deterministic Base64 Behavior

### LINKED ISSUE #333: Deterministic Base64 support
This issue is being created to supersede similarly reported issues and represent a canonical issue to enable the following:

1. Check to see if `java.util.Base64` is available at runtime using `Classes.isAvailable` (or similar), and if so, use that.  This is the [best/fastest one available](http://java-performance.info/base64-encoding-and-decoding-performance/) to most default environments and should be used whenever possible.
    
    `java.util.Base64` is available on [JDK >= 8](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html) and [Android level >= 26](https://developer.android.com/reference/java/util/Base64)

2. If `java.util.Base64` is not available:
    
    1. If the runtime enviornment is Android, use [android.util.Base64](https://developer.android.com/reference/android/util/Base64).  This is available on Android API level >= 8 (i.e. >= Android 2.2, "Froyo")
    
    2. Otherwise, assume < JDK8, and use `javax.xml.bind.DatatypeConverter`.  JDK 7 has this included automatically.

3. In any case, enable `JwtBuilder` and `JwtParser` methods that allow a user to specify their own Base64 `TextCodec` if/when desired, e.g. `Jwts.parser().setBase64Codec(myBase64Codec)`   This last option would allow the user to enable Base64 support in any environment that doesn't match 1 and 2 above.

4. Ensure any codec implementation will throw exceptions on invalid inputs when possible.  See #143 as an example.  Exceptions should be thrown when possible to indicate invalid input, wrapping/propagating the original codec exception.

Implementation note:  Perhaps instead of using `javax.xml.bind.DatatypeConverter` - which doesn't always fail on invalid inputs (see #143), we should embed the Commons Codec or MigBase64 implementation.

### PR DESCRIPTION
**Pull request by lhazlewood**
Fixes #333 

- Implemented new Base64 encoder forked from MigBase64 to guarantee deterministic behavior on all JDK and Android platforms
- Allowed pluggable Encoder/Decoder for JWT building and parsing via new `Encoder` and `Decoder` and `JwtBuilder#base64UrlEncodeWith` and `JwtParser#base64UrlDecodeWith` methods respectively
- Added tests for all new code to retain 100% code coverage
- Enabled `oraclejdk10` and `openjdk10` builds in TravisCI
- Replaced gmaven plugin with gmavenplus to work on JDK >= 9
- Upgraded surefire and failsafe plugins to 2.22.0 to ensure build works on JDK >= 10
- Ensured JavaDoc linter wouldn't fail the build for JDK >= 8 (was previously only 1.8)

### CODE REVIEW

**Comment by lhazlewood** [comment]
@RyanBard and @azagniotov - I'd appreciate a code review if you get a chance!

File: src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+class Base64DecoderTest {
+
+    @Test(expected = IllegalArgumentException)
+    void testDecodeWithNullArgument() {
+        new Base64Decoder().decode(null)
+    }
+
+    @Test
+    void testDecode() {
+        String encoded = 'SGVsbG8g5LiW55WM' // Hello 世界
```
Comment (RyanBard) [inline src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: I would add in the the RFC test vectors also (I'm not sure where the best place to put them would be among the various encode/decode tests in this PR): https://tools.ietf.org/html/rfc4648#page-12

```
   BASE64("") = ""

   BASE64("f") = "Zg=="

   BASE64("fo") = "Zm8="

   BASE64("foo") = "Zm9v"

   BASE64("foob") = "Zm9vYg=="

   BASE64("fooba") = "Zm9vYmE="

   BASE64("foobar") = "Zm9vYmFy"
```

It would also be nice to cook up some examples to distinguish the differences between the url-safe one also.

File: src/main/java/io/jsonwebtoken/JwtBuilder.java [hunk 2]
```diff
@@ -345,11 +347,36 @@
      * <p>This is a convenience method: the string argument is first BASE64-decoded to a byte array and this resulting
      * byte array is used to invoke {@link #signWith(SignatureAlgorithm, byte[])}.</p>
      *
+     * <h4>Deprecation Notice: Deprecated as of 0.10.0, will be removed in 1.0.0</h4>
```
Comment (RyanBard) [inline src/main/java/io/jsonwebtoken/JwtBuilder.java]: I notice the deprecation comments say 0.10.0, but this PR seems to be targeting 0.9.1-SNAPSHOT.

If this isn't a mistake, do you mind giving me a brief explanation of your branching/release process (high level enough that it's no more than 2-3 min of your time)?

File: src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy (hunk 1 above)
Comment (lhazlewood) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: Code coverage reports 100% of Base64's implementation covered, for all statements and branches.

That said, definitely happy to add as many test vectors as we can reliably trust :)

File: src/main/java/io/jsonwebtoken/JwtBuilder.java (hunk 2 above)
Comment (lhazlewood) [reply src/main/java/io/jsonwebtoken/JwtBuilder.java]: Ah, the PR targets master and master *should* be at `0.10.0-SNAPSHOT` not at `0.9.1-SNAPSHOT` as it currently is.  I need to change that - thanks for the catch.

Technically the deprecation notices themselves can target any patch release (0.9.1, 0.9.2, etc) because they're just JavaDoc changes and the signatures of all _non-internal/private_ API classes/methods are not changing.  That said, it was too much of a hassle to create a separate PR and release just for that so I'm bundling those notifications in the 0.10.0 release.

The other changes associated with this PR have to be 0.10.0 because [semver](https://semver.org/) requires a minor revision increase when new public APIs are introduced (e.g. `Encoder`, `Decoder`, new public methods on `JwtBuilder` and `JwtParser` etc).

So, based on semver, the way I manage all my git repos is:

`master` always reflects the next minor point release baseline (i.e. 'SNAPSHOT' in maven terminology) - 0.10.0-SNAPSHOT, 0.11.0-SNAPSHOT, 1.0.0-SNAPSHOT, etc) and contains all PRs/fixes/whatever pending the Maj.min.0 release.

If a bugfix or patch release is warranted for any `Maj.min.0` release, a branch is created from the `Maj.min.0` release tag with a `MAJ.min.x` naming scheme (suffix is the actual string literal `.x`)

For example, if 0.9.0 has been released, and we need to get a patch out (e.g. upgrading Jackson as I did earlier this week), I create a `0.9.x` branch from the `0.9.0` release tag. and that branch reflects `0.9.1-SNAPSHOT` and contains merged PRs awaiting release.

When 0.9.1 is about to be released, the pending changes also get merged to master (to ensure they're always in all future Maj.min releases).  Then 0.9.1 is released and then the 0.9.x branch reflects `0.9.2-SNAPSHOT` in its pom.xml.

At this point we can either keep the `0.9.x` branch around if we think there might be further patch releases (`0.9.2`, `0.9.3`, etc), or just delete it.  If we delete it and a `0.9.2` release is warranted, we would re-create the `0.9.x` branch based on the `0.9.1` tag.

I tend to prefer to delete `.x` branches after a patch release to keep things clean unless I know there is an immediate need for another follow-up patch very soon.

So in summary:

1. `master` always reflects the staged changes about to go out in the next `Maj.min.0` release and should always reflect `Maj.min.0-SNAPSHOT` in its pom.xml.

2. `Maj.min.x` branches (suffixed with the string literal '`x`') always reflect staged changes about to go out in the next Maj.min.patch release and should always reflect `Maj.min.patch-SNAPSHOT` in its pom.xml.

3. After cutting a `Maj.min.patch` release, those changes should also be merged into master (except for pom.xml changes that violate # 1).

I hope this helps!  Does that make sense?

File: src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy (hunk 1 above)
Comment (RyanBard) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: I wrote a quick program to generate printable ascii chars (it's a decent start at least, and we already have at least a few unicodes in the existing tests):

```
"special: [\n \t], ascii[32..126]: [ !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~]\n"
```

I manually tinkered with it a little, so if my base64 encoding is off a little, it's probably me.  We should also dump a \r in there (I don't trust myself making a \r in vim and didn't write this out to a file in code like I probably should have).

Linux's base64 of it:

```
c3BlY2lhbDogWwogXHRdLCBhc2NpaVszMi4uMTI2XTogWyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+XQo=
```

(again, subtle details might be wrong, so verify independent of my manual tinkering -- manual tinkering being where I added slashes b/c I printed the combination of ascii rather than writing to a file)

I'll write it to a file properly in a bit (going to eat lunch rq).

File: src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java [hunk 3]
```diff
@@ -294,25 +301,19 @@ public String compact() {
 
         String base64UrlEncodedHeader = base64UrlEncode(jwsHeader, "Unable to serialize header to json.");
 
-        String base64UrlEncodedBody;
+        byte[] bytes;
+        try {
+            bytes = this.payload != null ? payload.getBytes(Strings.UTF_8) : toJson(claims);
```
Comment (azagniotov) [inline src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java]: Should you also trim the `this.payload` and check if it is not empty before getting the bytes?

File: src/main/java/io/jsonwebtoken/JwtBuilder.java (hunk 2 above)
Comment (lhazlewood) [reply src/main/java/io/jsonwebtoken/JwtBuilder.java]: I just converted the previous reply to a wiki entry: https://github.com/jwtk/jjwt/wiki/Release-Code-Management for future reference.

**Comment by azagniotov** [comment]
I think the code changes look fine. I agree with @RyanBard to have more test vectors in place

File: src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java (hunk 3 above)
Comment (lhazlewood) [reply src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java]: Not really?  I mean, leading and trailing space could be a valid payload, and we don't want to sanitize people's payloads.  That said, this code isn't new - it's the same as it was before, just moved up a couple of lines, so I wouldn't want to change it unless there was a strong reason (e.g. in the RFC) to do so.  Or at the very least open a new issue to track it and only do it on a 1.0.0 release.

Comment (azagniotov) [reply src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java]: Alright, just wanted to confirm 👍

File: src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy (hunk 1 above)
Comment (lhazlewood) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: @RyanBard I added the suggested test vectors (and duplicated them for Base64Url too).  PR has been updated.

Comment (RyanBard) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: I'm reasonably confident in this one also:

```
input: "special: [\r\n \t], ascii[32..126]: [ !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~]\n"
encoded: "c3BlY2lhbDogWw0KIAldLCBhc2NpaVszMi4uMTI2XTogWyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+XQo="
```

I wrote that string to a file and used Linux's base64 util to encode (and then verified the decoding of that encoded str was my original output).

Comment (lhazlewood) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: @RyanBard in the block directly above, is everything to the right of `input: ` Java code? i.e. I can just copy-and-paste from that into a .java source code file?

Comment (RyanBard) [reply src/test/groovy/io/jsonwebtoken/codec/impl/Base64DecoderTest.groovy]: They were just string literals to dump into an assert, but since your tests are groovy and double quotes are GStrings, it required a little modification (and some replaces for the base64url).

I've PR'd into your branch here: https://github.com/jwtk/jjwt/pull/342

Scavenge what you need from it.

**Comment by lhazlewood** [comment]
Merged - many thanks to you @RyanBard and @azagniotov for your help and review on this one!
